### DISCUSSION TITLE
Tasmota and DTH22,AM2302...

### DISCUSSION THREAD

**Discussion by vidhome** [arendst/Tasmota#15697] [Ideas]
Gentlemen Tasmota experts and developers, very often there are questions about the operation of DTH22(AM2302) with Tasmota firmware on ESP8266 and ESP32. I myself have been fiddling with the problem of periodic non-reading of data in the ESP32 Pro from Wemos for a long time. I tried different connections to different outputs, pull-up resistors, etc. It turned out that the Tasmota default/DHT-sensor-library/DHT.cpp library needs to be corrected in case DHT22: delayMicroseconds(1100) value to delayMicroseconds(2000). Everything worked without failure. I propose to correct this parameter in the source codes.

**Comment by barbudor** [comment]
I don't have any of those crappy unreliable DHT sensors, so I couldn't validate that 2ms (like for DHT11) would be THE solution for everyone.
The [datasheet](https://www.sparkfun.com/datasheets/Sensors/Temperature/DHT22.pdf) requires only "above 1ms", and 2ms is obviously "above 1ms" so compliant.

If you have details showing that it would solve those problems for everyone, you are welcome to provide a PR

**Reply by vidhome** [reply]
Thanks for the answer, I can't test on all versions of ESP32, now I checked on version 11.1.0.3 on ESP32 and ESP8266, AM2302 works fine.  I agree that these are not the best sensors, it just so happens that I have a lot of them and are quite satisfied with the price and accuracy.  Although you probably need to migrate to I2C.

**Comment by arendst** [comment]
Not sure what Tasmota version you use but ESP8266 and ESP32 use different drivers for DHT sensors.

The ESP8266 uses adapted adafruit code with the requested timing change present since 20200621 (#7468).

The ESP32 uses the "standard" adafruit DHT library to get rid of user timing questions. It's not stable but then it's not a Tasmota fault anymore ;-). This library indeed is set to 1100 uSecs.

I'll do a local test changing it to 2000 and see if my esp32-c3 reports more stable results (without the DHT: Invalid reading messages). Hold on.....

**Comment by arendst** [comment]
ALas, changing it to 2000uS doesn't make it better:
```
13:36:26.613 DHT: Invalid reading
13:36:54.591 DHT: Invalid reading
13:37:10.614 DHT: Invalid reading
13:38:18.601 DHT: Invalid reading
13:39:10.586 DHT: Invalid reading
13:39:46.607 DHT: Invalid reading
13:39:54.589 DHT: Invalid reading
13:40:03.572 MQT: tele/esp32c3a/STATE = {"Time":"2022-05-26T13:40:03","Uptime":"0T00:05:09","UptimeSec":309,"Heap":202,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":36},"POWER":"OFF","Dimmer":50,"Color":"4040","White":50,"CT":326,"Channel":[25,25],"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"indebuurt_IoT","BSSId":"18:E8:29:CA:17:C1","Channel":11,"Mode":"11n","RSSI":100,"Signal":-23,"LinkCount":1,"Downtime":"0T00:00:03"}}
13:40:03.636 MQT: tele/esp32c3a/SENSOR = {"Time":"2022-05-26T13:40:03","DS18B20":{"Id":"01144A0CB2AA","Temperature":24.5},"AM2301":{"Temperature":24.6,"Humidity":23.8,"DewPoint":2.5},"ESP32":{"Temperature":41.8},"TempUnit":"C"}
13:40:26.625 DHT: Invalid reading
13:40:34.602 DHT: Invalid reading
13:40:42.581 DHT: Invalid reading
13:41:02.614 DHT: Invalid reading
13:41:34.615 DHT: Invalid reading
13:41:46.612 DHT: Invalid reading
13:42:22.612 DHT: Invalid reading
13:42:34.612 DHT: Invalid reading
13:42:38.574 DHT: Invalid reading
13:42:46.611 DHT: Invalid reading
13:42:54.595 DHT: Invalid reading
13:42:58.615 DHT: Invalid reading
13:43:02.577 DHT: Invalid reading
```

**Reply by vidhome** [reply]
Thanks for your answer, you are correct that ESP 8266 and ESP32 use different drivers.  I have an ESP32 not 32 S3 and it periodically (unpredictably) gives a read error.  now firmware from 2000 works on both ESP32 and ESP8266.  There are no problems, Tasmota version is 11.1.0.3.  That's exactly what he wrote.

**Comment by sfromis** [comment]
By now, I don't really see those 1-wire protocol sensors as something to waste my time on, when I2C sensors tend to just work. With various clones, and fickle timing requirements, finding one setup always working may be way too much trouble.

Arguably, the various timing factors in drivers could be moved to user_config build time settings (or even a command), to allow people to easier mess around with trying out what happens to work with their sensor and MCU.

**Reply by vidhome** [reply]
I agree, I2C is more reliable and accurate, DHT22 has a good price and a smaller number of wires for removal is more convenient.  It is very often used for cheap crafts.  It is possible to correctly make in user_config the ability to change the pulse durations .

**Reply by sfromis** [reply]
As arendst wrote
> The ESP32 uses the "standard" adafruit DHT library to get rid of user timing questions. It's not stable but then it's not a Tasmota fault anymore ;-)

If you want to play around.....
https://github.com/arendst/Tasmota/blob/development/lib/default/DHT-sensor-library/DHT.cpp#L261

**Comment by KoopsF** [comment]
If I understand correctly, support for the DTH22 is not coming anytime soon because it is difficult to communicate?

I own a DHT11 ESP-01 module and it appears to be working and communicating alright but the DHT22 ESP-01 does not at the moment.

**Reply by sfromis** [reply]
DHT22 is supported, but some users are having problems with some sensors.

**Reply by KoopsF** [reply]
I use the release sensors firmware and can only select the DHT11 from the menu, not the DHT22. When I try the DHT22 with the DHT11 setting I get the wrong measurements. I'm probably someone experiencing problems with my sensor like you said.

**Reply by sfromis** [reply]
[Select](https://tasmota.github.io/docs/Components/) AM2301 for the [support](https://tasmota.github.io/docs/AM2301/) of DHT21 DHT22 AM2301 AM2302 AM2321

**Reply by KoopsF** [reply]
Alright, yeah that seems to have worked. Thanks.
